src/Hpack/Config.hs

```diff
@@ -150,11 +153,14 @@ instance FromJSON (CaptureUnknownFields FlagSection) where
 
 getUnknownFields :: forall a. HasFieldNames a => Value -> Proxy a -> [FieldName]
 getUnknownFields v _ = case v of
-  Object o -> unknown
+  Object o -> ignoreUnderscored unknown
     where
       unknown = keys \\ fields
       keys = map T.unpack (HashMap.keys o)
       fields = fieldNames (Proxy :: Proxy a)
+      ignoreUnderscored
+        | warnUnderscoredUnknownFields (Proxy :: Proxy a) = id
+        | otherwise = filter (not . isPrefixOf "_")
```

Comment:
I like your code, it's quite elegant and extensible if we ever want to change the behavior in the future.  Yet, being advocate of TDD, I usually tend to go with the simplest solution that solves the problem and defer more clever solutions until we actually need them.

Further down we have a line

```haskell
mkPackage dir (CaptureUnknownFields unknownFields globalOptions@Section{sectionData = PackageConfig{..}}) = do
```

Here you could make the changes with a two liner (using view patterns):

```haskell
ignoreUnderscored = filter (not . isPrefixOf "_")

mkPackage dir (CaptureUnknownFields (ignoreUnderscored -> unknownFields) globalOptions@Section{sectionData = PackageConfig{..}}) = do
```